### DISCUSSION TITLE
OpenSSL FIPS provider SHA-1

### DISCUSSION THREAD

**Discussion by kgoldman** [openssl/openssl#25777] [Q&A]
I created a small test case loading the fips provider as per this code.

https://docs.openssl.org/master/man7/fips_module/#programmatically-loading-the-fips-module-nondefault-library-context

I sign a message using SHA-1.

    *md = EVP_get_digestbyname(str);
    irc = EVP_PKEY_CTX_set_signature_md(ctx, md);
    etc.

I expected this to fail because of SHA-1. I tried with and without
loading the "base" provider and they both succeeded.

What did I miss?

**Comment by t8m** [comment]
Can you please post the whole code chunk doing the signing?

**Reply by kgoldman** [reply]
The FIPS initialization was from the above openssl link. Using 'sha1' as the string:

	*rsaKey = PEM_read_PrivateKey(pemKeyFile, NULL, NULL, (void *)password);
	ctx = EVP_PKEY_CTX_new(rsaKey, NULL);	
	irc = EVP_PKEY_sign_init(ctx);
        *md = EVP_get_digestbyname(str);
        irc = EVP_PKEY_CTX_set_signature_md(ctx, md);
	irc = EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PADDING);
	irc = EVP_PKEY_sign(ctx, ...);

**Comment by t8m** [comment]
Also have you checked that you do not have the default provider loaded in the library context by any chance?

**Reply by kgoldman** [reply]
How would I do that?

**Reply by paulidale** [reply]
`OSSL_PROVIDER_available()` will do this.

What version of OpenSSL & FIPS provider are you using?

**Reply by kgoldman** [reply]
> `OSSL_PROVIDER_available()` will do this.

That is unexpected:

    irc = OSSL_PROVIDER_available(NULL, "fips");
returns 0, an error.

However

    fips = OSSL_PROVIDER_load(NULL, "fips");

returns non-NULL. Is that inconstant, it's not available, but it loads?

> 
> What version of OpenSSL & FIPS provider are you using?

OpenSSL 3.1.4 24 Oct 2023 (Library: OpenSSL 3.1.4 24 Oct 2023)

**Reply by paulidale** [reply]
A zero return from `OSSL_PROVIDER_available()` means the provider isn't available, it isn't an error return.

The FIPS provider doesn't magically load if you build OpenSSL with `enable-fips`.  You still have to load it either through configuration or programmatically.  Read the `README-FIPS.md` file and the relevant man pages for how to do this.

3.1.4 is also not FIPS validated and quite possibly never will be.  Again read the documentation about building one version of the FIPS provider and running it with a different version of OpenSSL.  Currently, the FIPS validated providers can only be built from 3.0.8 or 3.0.9 source code.

**Reply by kgoldman** [reply]
1. If 3.1.4 is not FIPS validated, what does it mean to load a FIPS provider?

2. It does seem surprising that the FIPS provider is not available and yet it can be loaded. There is /usr/lib64/ossl-modiles/fips.so

3. There is a note on the web site: OpenSSL version 3 contains a FIPS module, see [the FIPS module manual page for more information. This FIPS module is validated, and has the certificate number 4282.

**Reply by paulidale** [reply]
1. It means you are loading a non-validated FIPS provider and are not FIPS compliant.  Due to the way the validation process works, it is often a year or more after submission that the validation is achieved.  It's not possible to restrict the distribution to only validated providers without accurate precognition.
2. Why?  Unless you tell OpenSSL to load the FIPS provider, it doesn't happen.
3. Currently only OpenSSL versions 3.0.8 and 3.0.9 produce validated FIPS providers.  **All** other versions of OpenSSL do not.  This is documented on the web site.

**Reply by StephenWall** [reply]
As I recall, loading the approved 3.0.9 FIPS module is allowed in all later versions of OpenSSL.  So building the FIPS provider from 3.0.9 source following the security protocol, and using it with the (unapproved) 3.1.4 openssl libraries and command *is* operating in a validated FIPS mode.
